doc/design.md

```diff
@@ -274,3 +274,32 @@ Notes:
 
 - Store passwords encrypted
 - Use Postgres features for exporting JSON
+
+Database Functions
+--------------
+Functions that query or update the database will be organized into a single module in the expungeservice. 
+
+    create_user(email, admin, hashed_password)
+
+Insert a new user into the Users table with the given email string and admin flag, and inserts the password hash into the Auth table and link it with the User uuid. The uuid of each, and the date_created and date_updated are generated within the database.
+This pair of inserts is an atomic operation, so if one fails then the other one is cancelled and has no effect. 
+Returns None if insert is successful; otherwise throws an error.
+
+    check_user(email)
+
+Check the Users table to see if a user with the given email already exists. Returns boolean. If query fails, throws an error.
+
+    get_password_hash(email)
+
+Return the password_hash associated with an email. The password hash can then be compared against a password in the calling code. Assumes the user exists, and throws an error otherwise.
+
+Performing login is two steps: check the user exists, and if so then compare the password.
+User creation is also two steps: check the user doesn’t exist yet, and if not then create the user.
+
```

Comment:
What is the reason for making the user login and creation a two step process?

Reply:
That's true, and the more I think about it the more I like the idea of each being a single database operation that uses error-handling to parse the different outcomes. Arun and I talked about this briefly and were thinking that it might be simpler to break each down into the two pieces of logic. I forget if there were other reasons for this. @a-ilango , your thoughts?